### PR TITLE
copy_file: add rule and tests

### PR DESCRIPTION
**Pull request by laszlocsomor**
This PR adds two new rules: copy_file and
copy_xfile.

Both rules solve a common problem: to copy one
file to another location. The problem is routinely
solved using a genrule. That however requires
Bash, since genrules execute Bash commands.
Requiring Bash is a problem on Windows.

The new rules do not require Bash on Windows (only
on other platforms).

The only difference between the rules is that
copy_xfile creates an executable file while
copy_file doesn't.

See https://github.com/bazelbuild/bazel/issues/4319

### CODE REVIEW

File: rules/copy_file.bzl [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+"""
+
+load(
+    ":copy_file_private.bzl",
+    _copy_file = "copy_file",
+    _copy_xfile = "copy_xfile",
+)
+
+copy_file = _copy_file
+copy_xfile = _copy_xfile
```
Comment (thomasvl) [inline rules/copy_file.bzl]: Drive by - why create two rules vs. one rule with an argument to say if it should be executable?  seems like an `executable` or `make_executable` argument would be more self descriptive than _xfile_ as that's a bash-ism, no?

File: rules/copy_file_private.bzl [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    if ctx.attr.is_windows:
+        # Most Windows binaries built with MSVC use a certain argument quoting
+        # scheme. Bazel uses that scheme too to quote arguments. However,
+        # cmd.exe uses different semantics, so Bazel's quoting is wrong here.
+        # To fix that we write the command to a .bat file so no command line
+        # quoting or escaping is required.
+        bat = ctx.actions.declare_file(ctx.label.name + "-cmd.bat")
+        ctx.actions.write(
+            output = bat,
+            content = "@copy /Y \"%s\" \"%s\" >NUL" % (
```
Comment (thomasvl) [inline rules/copy_file_private.bzl]: lib/shell.bzl shell.escape on the args?

File: rules/copy_file.bzl (hunk 1 above)
Comment (laszlocsomor) [reply rules/copy_file.bzl]: See the same question here: https://github.com/bazelbuild/bazel-skylib/pull/122#issuecomment-470523889

**Comment by thomasvl** [comment]
And to keep @c-parsons happy, don't for get to update `rules/BUILD` and `docs/BUILD` so there is documentation for the new rules.  😄

File: rules/copy_file_private.bzl (hunk 2 above)
Comment (laszlocsomor) [reply rules/copy_file_private.bzl]: I see no such function: https://github.com/bazelbuild/bazel-skylib/blob/master/lib/shell.bzl

Which one do you mean?

Comment (thomasvl) [reply rules/copy_file_private.bzl]: Sorry, `shell.quote`

Comment (laszlocsomor) [reply rules/copy_file_private.bzl]: Unfortunately that uses Bash quoting syntax, which doesn't apply for cmd.exe

**Comment by laszlocsomor** [comment]
@thomasvl , thank you very much for the comments!

File: rules/copy_file_private.bzl [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            progress_message = "Copying files",
+            use_default_shell_env = True,
+        )
+    else:
+        ctx.actions.run_shell(
+            inputs = [ctx.file.src],
+            outputs = [ctx.outputs.out],
+            command = "cp -f \"$1\" \"$2\"",
+            arguments = [ctx.file.src.path, ctx.outputs.out.path],
+            mnemonic = "StarlarkCopyFile",
```
Comment (laurentlb) [inline rules/copy_file_private.bzl]: I don't think the word Starlark is useful here

File: rules/copy_file.bzl (hunk 1 above)
Comment (laurentlb) [reply rules/copy_file.bzl]: Making a rule executable allows the user to do `bazel run`. Is it useful here?

File: rules/copy_file_private.bzl [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        **kwargs
+    )
+
+def copy_file(name, src, out, **kwargs):
+    _copy_file(
+        name = name,
+        src = src,
+        out = out,
+        is_windows = select({
+            "@bazel_tools//src/conditions:host_windows": True,
```
Comment (laurentlb) [inline rules/copy_file_private.bzl]: Where is this label documented?

File: rules/copy_file.bzl (hunk 1 above)
Comment (laszlocsomor) [reply rules/copy_file.bzl]: Yes, but more than that: it allows putting the rule into sh_binary.srcs and py_test.srcs.

File: rules/copy_file_private.bzl (hunk 4 above)
Comment (laszlocsomor) [reply rules/copy_file_private.bzl]: I don't know if it's documented at all.

File: rules/copy_file_private.bzl (hunk 3 above)
Comment (laszlocsomor) [reply rules/copy_file_private.bzl]: Removed.

**Comment by laszlocsomor** [comment]
```
🛠️ Printing raw output for debugging | 0s
-- | --
  | ##### :bazel: buildifier: error while downloading Buildifier:
  | <pre><code>Unknown Buildifier version &#x27;True&#x27;</code></pre>
```

Looks like an error in Buildifier or the Buildkite pipeline?

**Comment by laszlocsomor** [comment]
> documentation for the new rules. 😄

Thanks for the advice! I tried doing that, but I can't build any targets from under `//docs`, so I can't verify if what I did is correct. So I'm not updating any commits for now.

**Comment by thomasvl** [comment]
> Thanks for the advice! I tried doing that, but I can't build any targets under //docs, so I can't verify if what I did is correct. So I'm not uploading any commits for now.

Did you notice: http://go/gh/bazelbuild/bazel-skylib/blob/master/docs/BUILD#L3-L4

**Comment by laszlocsomor** [comment]
Ah, of course I didn't. Thanks. Uploaded the new commit.

Below is the generated documentation.

---

## write_file

<pre>
write_file(<a href="#write_file-name">name</a>, <a href="#write_file-out">out</a>, <a href="#write_file-content">content</a>, <a href="#write_file-is_executable">is_executable</a>, <a href="#write_file-kwargs">kwargs</a>)
</pre>

Creates a UTF-8 encoded text file.

### Parameters

<table class="params-table">
  <colgroup>
    <col class="col-param" />
    <col class="col-description" />
  </colgroup>
  <tbody>
    <tr id="write_file-name">
      <td><code>name</code></td>
      <td>
        required.
        <p>
          Name of the rule.
        </p>
      </td>
    </tr>
    <tr id="write_file-out">
      <td><code>out</code></td>
      <td>
        required.
        <p>
          Path of the output file, relative to this package.
        </p>
      </td>
    </tr>
    <tr id="write_file-content">
      <td><code>content</code></td>
      <td>
        optional. default is <code>[]</code>
        <p>
          Lines of text, the contents of the file. Newlines are added
  automatically after every line except the last one.
        </p>
      </td>
    </tr>
    <tr id="write_file-is_executable">
      <td><code>is_executable</code></td>
      <td>
        optional. default is <code>False</code>
      </td>
    </tr>
    <tr id="write_file-kwargs">
      <td><code>kwargs</code></td>
      <td>
        optional.
      </td>
    </tr>
  </tbody>
</table>

**Comment by laszlocsomor** [comment]
Here's the rendering of the documentation.

---


## copy_file

<pre>
copy_file(<a href="#copy_file-name">name</a>, <a href="#copy_file-src">src</a>, <a href="#copy_file-out">out</a>, <a href="#copy_file-is_executable">is_executable</a>, <a href="#copy_file-kwargs">kwargs</a>)
</pre>

Copies a file to another location.

`native.genrule()` is sometimes used to copy files (often wishing to rename them). The 'copy_file' rule does this with a simpler interface than genrule.

This rule uses a Bash command on Linux/macOS/non-Windows, and a cmd.exe command on Windows (no Bash is required).


### Parameters

<table class="params-table">
  <colgroup>
    <col class="col-param" />
    <col class="col-description" />
  </colgroup>
  <tbody>
    <tr id="copy_file-name">
      <td><code>name</code></td>
      <td>
        required.
        <p>
          Name of the rule.
        </p>
      </td>
    </tr>
    <tr id="copy_file-src">
      <td><code>src</code></td>
      <td>
        required.
        <p>
          A Label. The file to make a copy of. (Can also be the label of a rule
    that generates a file.)
        </p>
      </td>
    </tr>
    <tr id="copy_file-out">
      <td><code>out</code></td>
      <td>
        required.
        <p>
          Path of the output file, relative to this package.
        </p>
      </td>
    </tr>
    <tr id="copy_file-is_executable">
      <td><code>is_executable</code></td>
      <td>
        optional. default is <code>False</code>
        <p>
          A boolean. Whether to make the output file executable. When
    True, the rule's output can be executed using `bazel run` and can be
    in the srcs of binary and test rules that require executable sources.
        </p>
      </td>
    </tr>
    <tr id="copy_file-kwargs">
      <td><code>kwargs</code></td>
      <td>
        optional.
      </td>
    </tr>
  </tbody>
</table>

File: rules/copy_file_private.bzl (hunk 2 above)
Comment (c-parsons) [reply rules/copy_file_private.bzl]: Perhaps a comment to that effect?

File: tests/copy_file/BUILD [hunk 5]
```diff
@@ -0,0 +1,86 @@
+load("//rules:copy_file.bzl", "copy_file")
```
Comment (c-parsons) [inline tests/copy_file/BUILD]: (same comment as your other PR: I find the overall goal and structure of this BUILD file unclear. Can you add a top-level comment describing it?)

File: docs/BUILD [hunk 6]
```diff
@@ -4,98 +4,104 @@ load("@io_bazel_skydoc//stardoc:stardoc.bzl", "stardoc")
 # the --incompatible_remap_main_repo flag.
 stardoc(
     name = "build_test_docs",
-    input = "//rules:build_test.bzl",
     out = "build_test_doc_gen.md",
+    input = "//rules:build_test.bzl",
```
Comment (c-parsons) [inline docs/BUILD]: pull and merge

File: tests/copy_file/BUILD [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+package(default_testonly = 1)
+
+sh_test(
+    name = "copy_file_tests",
+    srcs = ["copy_file_tests.sh"],
+    data = [
+        ":run_executables",
+        # Use DefaultInfo.files from 'copy_src' (via 'file_deps').
+        ":file_deps",
```
Comment (c-parsons) [inline tests/copy_file/BUILD]: buildifier

File: tests/copy_file/BUILD (hunk 5 above)
Comment (laszlocsomor) [reply tests/copy_file/BUILD]: Done.

File: docs/BUILD (hunk 6 above)
Comment (laszlocsomor) [reply docs/BUILD]: Done.

File: tests/copy_file/BUILD (hunk 7 above)
Comment (laszlocsomor) [reply tests/copy_file/BUILD]: Done.

File: rules/copy_file_private.bzl (hunk 2 above)
Comment (laszlocsomor) [reply rules/copy_file_private.bzl]: Done.

**Comment by laszlocsomor** [comment]
@laurentlb @brandjon @vladmos : could one of you merge this please?

Would it make sense to grant me merge rights? I'm planning to do more changes in this repo.

**Comment by laszlocsomor** [comment]
ping -- please help merging

**Pull request merged**